backend/OKPluginVariableRendering/VariableRendering.cs

```diff
@@ -0,0 +1,342 @@
+
+            //TODO: select only the realtions that are defined in configuration
+            //      check if selection with specific predicates is possible
+            var relations = new List<string>();
+
+            cfg.BaseCI.FollowRelations.ForEach(r =>
+            {
+                r.Follow.ForEach(f =>
+                {
+                    if (relations.IndexOf(f.Predicate[1..]) == -1)
+                    {
+                        relations.Add(f.Predicate[1..]);
+                    }
+                });
+            });
+
+            var allRelations = new List<Relation>();
+
+            foreach (var layer in cfg.InputLayerSet)
+            {
+                var layerRelations = await relationModel.GetRelations(RelationSelectionAll.Instance, layer, trans, changesetProxy.TimeThreshold);
+
+                allRelations.AddRange(layerRelations);
+            }
+
+            var fragments = new List<BulkCIAttributeDataLayerScope.Fragment>();
+
+            foreach (var mainCI in mainCIs)
+            {
+
+                var gatheredAttributes = new Dictionary<string, GatheredAttribute>();
+
+                foreach (var mapping in cfg.BaseCI.AttributeMapping)
+                {
+                    foreach (var (name, attribute) in mainCI.MergedAttributes)
+                    {
+                        if (IsAttributeIncludedInSource(name, mapping.Source))
+                        {
+                            var a = new GatheredAttribute
+                            {
+                                Name = GetTargetName(name, mapping.Target),
+                                Value = attribute.Attribute.Value,
+                                Priority = baseCIPrio,
+                            };
+
+                            if (gatheredAttributes.ContainsKey(a.Name))
+                            {
+                                // check priority of attributes
+                                if (gatheredAttributes[a.Name].Priority < a.Priority)
+                                {
+                                    gatheredAttributes[a.Name] = a;
+                                }
+                            }
+                            else
+                            {
+                                gatheredAttributes.Add(a.Name, a);
+                            }
+                        }
+                    }
+                }
+
+
+                foreach (var followRelation in cfg.BaseCI.FollowRelations)
+                {
+                    var prevCI = mainCI;
+                    foreach (var follow in followRelation.Follow)
+                    {
+                        // get the predicate remove the first char, first char defines the direction of relation
+                        var predicate = follow.Predicate[1..];
+
+                        // check relation direction
+                        var r = allRelations.Where(r =>
```

Comment:
Searching through a list using `.Where()` and `.FirstOrDefault()` is really slow and AFAICT a big reason why your implementation is slow as a whole. Because for each mainCI, you are going through ALL relations, as `.Where()` needs to do a linear search.

What I would do instead is create two dictionaries of relations (instead of the `allRelations` list) where one has the compound key predicate+fromCIID and the second has predicate+toCIID.

That way, you can do a O(1) lookup here instead of the O(N) search you have currently.